lib/internal/util/inspect.js

```diff
@@ -185,9 +186,13 @@ function inspect(value, opts) {
     if (typeof opts === 'boolean') {
       ctx.showHidden = opts;
     } else if (opts) {
-      const optKeys = Object.keys(opts);
-      for (var i = 0; i < optKeys.length; i++) {
-        ctx[optKeys[i]] = opts[optKeys[i]];
+      for (const key of Object.keys(opts)) {
+        if (inspectDefaultOptions[key] === undefined &&
+            key !== 'stylize' &&
+            !(key in inspectDefaultOptions)) {
+          throw new ERR_INVALID_OPT_KEY(key);
```

Comment:
This particular change seems unrelated, and as far as I can tell, it would mean that when we introduce a new option in the future, that would break on older versions of Node.js as opposed to being silently ignored? If my understanding is correct, I’m -1 on this.

Reply:
Without this, the `seen` array could be overridden (just as well as the `indentationLvl` and `budget`). That's not intended. I could also prevent this otherwise but this seemed the most straight forward way and I believe it will also uncover bugs in user implementation with typos.

Therefore I consider this something pretty important.

Using "newer" options in older Node.js versions that have this validation would indeed throw but I don't think this is bad at all. IMO if code is written for a specific Node.js version it should be upwards compatible but not downwards compatible.

Every userland module normally supports a version range from a specific point on (e.g., `=> 6.x`). That's just the same here: it's a feature that can only be used for new enough Node.js versions (it would be `=> 12.x`).

Reply:
> Therefore I consider this something pretty important.

Then maybe we could do something like you did below, to filter out specific properties that we care about? Or would that be too slow?

> IMO if code is written for a specific Node.js version it should be upwards compatible but not downwards compatible.

Yes, I think that’s the part that I disagree with – this PR would mean that we *artificially* reduce the range of supported versions…

@nodejs/tsc Any opinions? I don’t want this to be blocked just on me.

Reply:
> Then maybe we could do something like you did below, to filter out specific properties that we care about? Or would that be too slow?

This would prevent the main aspect which I would like to get with it: finding bugs in user implementations by finding typos. Otherwise they would be silently ignored as well. The performance is not that critical here.

> this PR would mean that we artificially reduce the range of supported version

I don't think that it's artificially reducing the supported version range. Either something fully works in a version or not. Silently ignoring something could result in unwanted behavior. And isn't this what semver-major is for?
I consider it a breaking change feature.

Reply:
After gathering the general feedback that it's probably to late to change this in general, I am uncertain how to address this properly...

I see the following options:

1) Stop accepting only specific options. This would be inconsistent and seems "random": any other property will be passed through to the user in the custom inspection function.
2) Split the user options and internal options completely from each other. This requires a significantly bigger code change and an extra argument to be passed around in all functions calls that have something to do with `util.inspect`. There would also be a performance penalty in case we do not pass through the original options argument from the user but a new options object which is a clone from the current one without the internal properties and even without this, there might be a performance penalty.
3) For this specific API we keep my implementation and forbid all unknown options but this is an exception which would only apply to `util.inspect()`.

@addaleax @joyeecheung does either of you have another idea or suggestion?